File: 0. 알고리즘 개념공부/2주차/07. 너비 우선 탐색 예제/당근마켓.swift
//
//  당근마켓.swift
//  Algorithm
//
//  Created by 김동현 on 3/26/25.
//

import Foundation

struct Queue<T> {
    private var enqueueStack: [T] = []
    private var dequeueStack: [T] = []
    
    var size: Int {
        return enqueueStack.count + dequeueStack.count
    }
    
    var isEmpty: Bool {
        return enqueueStack.isEmpty && dequeueStack.isEmpty
    }
    
    mutating func enqueue(_ element: T) {
        enqueueStack.append(element)
    }
    
    @discardableResult
    mutating func dequeue() -> T? {
        // dequeueStack가 비어있다면, enqueueStack의 모든 요소를 뒤집어 deqeuueStack로 옮긴다
        if dequeueStack.isEmpty {
            dequeueStack = enqueueStack.reversed()
            enqueueStack.removeAll()
        }
        return dequeueStack.popLast()
    }
}

@main
struct Main {
    static func main() {
        freopen("input.txt", "r", stdin)
        
        let dy: [Int] = [-1, 0, 1, 0]
        let dx: [Int] = [0, 1, 0, -1]

        var input = readLine()!.split(separator: " ").map { Int($0)! }
        let n = input[0]
        let m = input[1]
        
        input = readLine()!.split(separator: " ").map { Int($0)! }
        let sy = input[0]
        let sx = input[1]
        
        input = readLine()!.split(separator: " ").map { Int($0)! }
        let ey = input[0]
        let ex = input[1]
        
        // 격자
        //var a: [[Int]] = []
        var a: [[Int]] =  Array(repeating: Array(repeating: 0, count: m), count: n)
        
        // 방문 배열
        var visited: [[Int]] = Array(repeating: Array(repeating: 0, count: m), count: n)
        
        for i in 0..<n {
            let row = readLine()!.split(separator: " ").map { Int($0)! }
            //a.append(row)
            a[i] = row
        }
        
        var queue = Queue<(Int, Int)>()
        visited[sy][sx] = 1
        queue.enqueue((sy, sx))
        
        while let (y, x) = queue.dequeue() {
            for i in 0..<4 {
                let ny = y + dy[i]
                let nx = x + dx[i]
                
                if ny<0 || ny>=n || nx<0 || nx>=m || a[ny][nx] == 0 {
                    continue
                }
                
                // 값이 0이 아닐경우
                if visited[ny][nx] != 0 { continue }
                
                visited[ny][nx] = visited[y][x] + 1
                queue.enqueue((ny, nx))
            }
        }
        print(visited[ey][ex])
        
    }
}

File: 0. 알고리즘 개념공부/2주차/07. 너비 우선 탐색 예제/당근마켓.cpp
//
//  2.cpp
//  Algorithm
//
//  Created by 김동현 on 3/26/25.
//

/*
 가중치가 같은 그래프 내의 최단거리 알고리즘
 */

#include <iostream>
#include <queue>
using namespace std;

const int max_n = 104;
int dy[4] = {-1, 0, 1, 0};
int dx[4] = {0, 1, 0, -1};
int n, m, a[max_n][max_n], visited[max_n][max_n], y, x, sy, sx, ey, ex;

int main() {
    freopen("input.txt", "r", stdin);
    cin >> n >> m;      // map 크기
    cin >> sy >> sx;    // 승원이 좌표
    cin >> ey >> ex;    // 당근 좌표
    
    for (int i=0; i<n; i++) {
        for (int j=0; j<m; j++) {
            cin >> a[i][j];
        }
    }
    
    queue<pair<int, int>> q;
    
    q.push({sy, sx});
    
    while(q.size()) {
        tie(y, x) = q.front(); q.pop();
        for (int i=0; i<4; i++) {
            int ny = y + dy[i];
            int nx = x + dx[i];
            // 갈수 없는곳 무시
            if (ny<0 || ny>=n || nx<0 || nx>=m || a[ny][nx] == 0) {
                continue;
            }
            // 방문하였다면 무시
            if (visited[ny][nx]) continue;
            
            // 진행
            visited[ny][nx] = visited[y][x] + 1;
            q.push({ny, nx});
            
        }
    }
    
    cout << visited[ey][ex] << "\n";
    for (int i=0; i<n; i++) {
        for (int j=0; j<m; j++) {
            cout << visited[i][j] << " ";
        }
        cout << "\n";
    }
    
    
    return 0;
}
